pytest_trio/_tests/test_trio_mode.py

```diff
@@ -36,3 +36,130 @@ def test_trio_mode(testdir, enable_trio_mode):
 
     result = testdir.runpytest()
     result.assert_outcomes(passed=2, failed=2)
+
+
+qtrio_text = """
+import trio
+
+fake_used = False
+
+def run(*args, **kwargs):
+    global fake_used
+    fake_used = True
+
+    return trio.run(*args, **kwargs)
+"""
+
+
+def test_trio_mode_and_qtrio_run_configuration(testdir):
+    testdir.makefile(
+        ".ini", pytest="[pytest]\ntrio_mode = true\ntrio_run = qtrio\n"
+    )
+
+    testdir.makepyfile(qtrio=qtrio_text)
+
+    test_text = """
+    import qtrio
+    import trio
+
+    async def test_fake_qtrio_used():
+        await trio.sleep(0)
+        assert qtrio.fake_used
+    """
+    testdir.makepyfile(test_text)
+
+    result = testdir.runpytest()
+    result.assert_outcomes(passed=1)
+
+
+def test_trio_mode_and_qtrio_marker(testdir):
+    testdir.makefile(".ini", pytest="[pytest]\ntrio_mode = true\n")
+
+    testdir.makepyfile(qtrio=qtrio_text)
+
+    test_text = """
+    import pytest
+    import qtrio
+    import trio
+
+    @pytest.mark.trio(run=qtrio.run)
+    async def test_fake_qtrio_used():
+        await trio.sleep(0)
+        assert qtrio.fake_used
+    """
+    testdir.makepyfile(test_text)
+
+    result = testdir.runpytest()
+    result.assert_outcomes(passed=1)
+
+
+def test_qtrio_just_run_configuration(testdir):
+    testdir.makefile(".ini", pytest="[pytest]\ntrio_run = qtrio\n")
+
+    testdir.makepyfile(qtrio=qtrio_text)
+
+    test_text = """
+    import pytest
+    import qtrio
+    import trio
+
+    @pytest.mark.trio
+    async def test_fake_qtrio_used():
+        await trio.sleep(0)
+        assert qtrio.fake_used
+    """
+    testdir.makepyfile(test_text)
+
+    result = testdir.runpytest()
+    result.assert_outcomes(passed=1)
+
+
+def test_invalid_trio_run_fails(testdir):
+    run_name = "invalid_trio_run"
+
+    testdir.makefile(
+        ".ini", pytest=f"[pytest]\ntrio_mode = true\ntrio_run = {run_name}\n"
+    )
+
+    test_text = """
+    async def test():
+        pass
+    """
+    testdir.makepyfile(test_text)
+
+    result = testdir.runpytest()
+    result.assert_outcomes()
+    result.stdout.fnmatch_lines(
+        [
+            f"*ValueError: {run_name!r} not valid for 'trio_run' config.  Must be one of: *"
+        ]
+    )
+
+
+def test_third_party_run_plus_marker_fails(testdir):
```

Comment:
Is there any reason not to make this case prefer the marker on the test? That seems like it should take precedence because it's more specific (smaller scope) than the marker in pytest.ini.

Reply:
Mostly I find it weird that no runner in config file + qtrio runner on the test works fine, while explicitly specifying the default trio runner in config file + qtrio runner on the test raises an error.

I agree that there's a theoretical potential for confusion around fixtures, but I'm not worried about it in practice: I think autouse async fixtures are pretty rare, and I'd be especially surprised to see an autouse qt-specific async fixture. And non-autouse fixtures are listed on the particular test, so if they're not compatible with that test's runner, I think it's the fault of whoever wrote the test. :-)

Another example: what if one wants to pass arguments to the runner for a particular test? You can do `@pytest.mark.trio(run=partial(qtrio.run, special=42))` but then if you said `trio_run = qtrio` in the config file that'll be a (spurious IMO) error.

I don't think we need detailed compatibility checks. Probably is worth a brief mention in the docs though.

Reply:
I agree that an unspecified `trio_run` shouldn't differ from explicitly specifying the default of `"trio"`.  But, I think the code is unaware of which case it is.

https://github.com/python-trio/pytest-trio/blob/c998963007df7ed49f1f927fe6e68854e4b28c87/pytest_trio/plugin.py#L44-L48

Regardless, the accuracy of your first scenario isn't relevant if we choose to take the approach of not checking compatibility at all.  I'm fine with changing that, at least for now, and will hopefully get it done today.

Reply:
Removed in https://github.com/python-trio/pytest-trio/pull/105/commits/68860e0cc703856e3dfa5600f5a04b39e2422b0b.